Lib/test/test_shelve.py

```diff
@@ -165,6 +166,277 @@ def test_default_protocol(self):
+
+        with self.assertRaises(dbm_sqlite3.error):
+            def serializer(obj, protocol=None):
+                pass
+
+            def deserializer(data):
+                return data.decode("utf-8")
+
+            with shelve.open(self.fn, serializer=serializer,
+                             deserializer=deserializer) as s:
+                s["foo"] = "bar"
+
+        def serializer(obj, protocol=None):
+            return type(obj).__name__.encode("utf-8")
+
+        def deserializer(data):
+            pass
+
+        with shelve.open(self.fn, serializer=serializer,
+                         deserializer=deserializer) as s:
+            s["foo"] = "bar"
+            self.assertNotEqual(s["foo"], "bar")
+            self.assertIsNone(s["foo"])
+
+    def test_custom_serializer_and_deserializer_bsd_db_shelf(self):
+        berkeleydb = import_helper.import_module("berkeleydb")
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        def serializer(obj, protocol=None):
+            data = obj.__class__.__name__
+            if protocol == 5:
+                data = str(len(data))
+            return data.encode("utf-8")
+
+        def deserializer(data):
+            return data.decode("utf-8")
+
+        for proto in range(pickle.HIGHEST_PROTOCOL + 1):
+            with self.subTest(proto=5):
+                with shelve.BsdDbShelf(
+                    berkeleydb.btopen(self.fn),
+                    protocol=proto,
+                    serializer=serializer,
+                    deserializer=deserializer,
+                ) as s:
+                    bar = "bar"
+                    bytes_data = b"Hello, world!"
+                    bytearray_data = bytearray(b"\x00\x01\x02\x03\x04")
+                    array_data = array.array("i", [1, 2, 3, 4, 5])
+
+                    s["foo"] = bar
+                    s["bytes_data"] = bytes_data
+                    s["bytearray_data"] = bytearray_data
+                    s["array_data"] = array_data
+
+                    if proto == 5:
+                        self.assertEqual(
+                            s["foo"], f"{len(type(bar).__name__)}"
+                        )
+                        self.assertEqual(
+                            s["bytes_data"],
+                            f"{len(type(bytes_data).__name__)}",
+                        )
+                        self.assertEqual(
+                            s["bytearray_data"],
+                            f"{len(type(bytearray_data).__name__)}",
+                        )
+                        self.assertEqual(
+                            s["array_data"],
+                            f"{len(type(array_data).__name__)}",
+                        )
+
+                        key, value = s.set_location(b"foo")
+                        self.assertEqual("foo", key)
+                        self.assertEqual(value, f"{len(type(bar).__name__)}")
```

Comment:
```suggestion
                        key, value = s.set_location(b"foo")
                        self.assertEqual(key, "foo")
                        self.assertEqual(value, f"{len(type(bar).__name__)}")
```

Let's assert key and value as LHS. And to save a few strokes, you can use 'k, v' instead Maybe The subsequent lines could be on a single line, namely instead of

```py
                        self.assertEqual(
                            value, f"{len(type(bytes_data).__name__)}"
                        )
```

we would be able to have

```py
                    self.assertEqual(v, typename_len(bytes_data))
```